distribution.pm

```diff
@@ -109,7 +110,12 @@ sub script_run {
     testapi::type_string "$cmd";
     if ($wait > 0) {
         my $str = testapi::hashed_string("SR$cmd$wait");
-        testapi::type_string " ; echo $str-\$?- > /dev/$testapi::serialdev\n";
+        if (testapi::is_serial_terminal) {
+            testapi::type_string " ; echo $str-\$?-\n";
+        }
+        else {
+            testapi::type_string " ; echo $str-\$?- > /dev/$testapi::serialdev\n";
+        }
```

Comment:
This is a good point, I will think about it, there is no way of the distribution or backend overriding some platform specific code in the testapi in the first place so it should be refactored anyway. I don't see a major conceptual problem with doing both input and output to a serial terminal in os-autoinst, it is just another IO method that isn't too different from the current VNC-serial-port hybrid method. In this particular case we just need to provide an abstraction for a "command delimiter".

Having said that I think it is OK to have an if statement with alternatives for graphical and text terminals because it is very unlikely there will be a third or forth type for the foreseeable future, but the testapi should not have any shell specific code in it.